### ISSUE TITLE
empty array initializer

### ISSUE THREAD

**Issue by vzaliva** [vellvm/vellvm#106]
The following construct:

```
@D = internal constant double [], align 16
```

Triggers "Undefined Behaviour" error when interpreted. Clang gives more reasonable error message:

```
$ clang test_pointwise_plusD.ll                               12:22:39
test_pointwise_plusD.ll:2:31: error: invalid empty array initializer
@D = internal constant double [], align 16
                              ^
1 error generated.
```

I think this is an error, not undefined behavior. LLVM language spec says: _"the array type requires a size (number of elements) and an underlying data type."_

It should probably be handled early, by the parser.

**Comment by YaZko** [comment]
Right, the incorrect qualification of Errors into Undefined Behaviour was fixed by 4f36a3820e493b0bdd13124a2789417cd9cc4067.

Is the current error message more satisfying?

**Comment by vzaliva** [comment]
Interestingly it no longer triggers any error. 

```
 $ llc -relocation-model=pic -asm-verbose -O0 test_pointwise_plusD.ll 

llc: test_pointwise_plusD.ll:2:31: error: invalid empty array initializer
@D = internal constant double [], align 16
```

I am attaching the full source code. Clang on the same file gives the following error:

```
 $ llc -relocation-model=pic -asm-verbose -O0 test_pointwise_plusD.ll 

llc: test_pointwise_plusD.ll:2:31: error: invalid empty array initializer
@D = internal constant double [], align 16
```
[test_pointwise_plusD.ll.txt](https://github.com/vellvm/vellvm/files/3735642/test_pointwise_plusD.ll.txt)

**Comment by YaZko** [comment]
Right our parser is definitely a bit too lenient on this one, I'll look into fixing this when I can.
The fact that it did not crash during the execution had be a bit confused, but it actually make some sense: you never look up the content of the array in your program, you just plays with its address.

**Comment by vzaliva** [comment]
But I do read it:

```
  %l4 = load double, double* @D, align 8
```

**Comment by YaZko** [comment]
Yes definitely sorry for my previous comment. I believe that the story is the following:
* Since the parser do not complain, we allocate an empty block in memory place of the empty array.
* During the load, since we load at a type whose size is strictly bigger than the size of the block, all extra-bytes are filled with undef bytes, which result in the load returning undef.
* We would expect UB to be raised when the store to this undef location occurs, which would be the answer from the model: in order to store at the location, it checks that the value concretizes to a singleton. However the interpreter disregards this checks, simply pick 0 as a concretization, and continue the execution.
The issue #111 discusses this. I believe it is a _sound_ behavior of the interpreter, anything being a fair refinement of UB, but arguably counter-intuitive.

**Comment by vzaliva** [comment]
It sounds like a sound behavior of the **interpreter**, but since the original language construct is invalid is should not pass the **parser**. So I will keep this issue open until we fix the parser. 

P.S. It is not urgent/critical

**Comment by YaZko** [comment]
I will close this issue as stale, as a lot has changed since then. Please reopen if it is still relevant and existing.
